### PR TITLE
Refactor and fixup graph loading

### PR DESCRIPTION
**Pull request by mstv**
Alternative to #10680

## Proposed changes

- Refactoring
  * `RevisionGraph.GetSegmentsForRow`: Handle negative `row`
  * Make `UpdateVisibleRowRangeInternalAsync` clearer
  * Make `StraightenLanes` clearer (taken from #10778)
- Fixes
  * Fixup incomplete condition for graph cache loaded
  * Avoid unnecessary and incomplete lookahead for graph straightening
  * Fixup endless graph loading loop
- Feature
  * Preload next graph page

## Screenshots 

N/A

## Test methodology 

- manual

## Merge strategy

Please do *not squash merge* the commits!

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs [hunk 1]
```diff
@@ -414,7 +416,9 @@ private void BuildOrderedRowCache(RevisionGraphRevision[] orderedNodesCache, int
                 localOrderedRowCache.Add(new RevisionGraphRow(revision, segments));
             }
 
-            StraightenLanes(startIndex - _straightenLanesLookAhead, lastToCacheRowIndex, localOrderedRowCache);
+            int straightenStartIndex = Math.Max(1, startIndex - _straightenLanesLookAhead);
```
Comment (gerhardol) [inline GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs:419]: why start with 1? (0?)

**Review by gerhardol — COMMENTED** [review]
+0
Proper review take a long time, seem reasonable though.

**Comment by RussKie** [comment]
This type has some great comments, but I feel like additional inline comments for the changed lines could further help. E.g., I find this very useful:
https://github.com/gitextensions/gitextensions/blob/de93fd8abded89ea0abe87229063e84b9651f087/GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs#L428-L448

A specific example: in lines:100-101 we're now multiplying by 2 - why and what does it mean?

File: GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs (hunk 1 above)
Comment (mstv) [reply GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs:419]: > why start with 1?

The single node at index 0 shall never be moved.
Starting with 1, we always have a previous row to compare with.
As Igor suggested, I am adding a few comments.

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs [hunk 2]
```diff
@@ -682,16 +683,19 @@ private async Task UpdateVisibleRowRangeInternalAsync()
 
                         if (AppSettings.ShowRevisionGridGraphColumn)
                         {
-                            int scrollTo;
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:685]: a pointless copy of `newBackgroundScrollTo`

**Review by RussKie — APPROVED** [review]

**Review by gerhardol — APPROVED** [review]

**Pull request merged**